src/notebook/notebookFeature.ts

```diff
@@ -49,11 +69,39 @@ export class JuliaNotebookFeature {
                 }
             })
             controller.executeHandler = this.executeCells.bind(this)
+            controller.onDidChangeSelectedNotebooks(({ notebook, selected }) => {
+                // If we select our controller, then update the notebook metadata with the kernel information.
+                if (!selected) {
+                    return;
+                }
+                this.updateNotebookWithSelectedKernel(notebook, displayName, ver);
+            }, this, this.disposables)
 
-            this._controllers.push(controller)
+            this._controllers.set(controller, {version: ver })
         }
     }
 
+    private onDidOpenNotebookDocument(e: vscode.NotebookDocument) {
+        if (!this.isJuliaNotebook(e) || this._controllers.size === 0) {
+            return;
+        }
+        // Get metadata from notebook (to get an hint of what version of julia is used)
+        const { name, version } = this.getKernelSpecNameAndVersion(e);
+        let preferredControllerFound = false;
+        this._controllers.forEach((info, controller) => {
+            // If we find a controller that matches the vesion in the notebook metadata, then set
+            // that controller as the preferred controller.
+            if (name.includes(info.version) || version.includes(info.version)) {
+                preferredControllerFound = true;
+                controller.updateNotebookAffinity(e, vscode.NotebookControllerAffinity.Preferred)
+            }
+        })
+        if (!preferredControllerFound) {
+            // We know its a Julia notebook, hence give preference to one of our controllers.
+            const preferredController = Array.from(this._controllers.keys())[0];
+            preferredController.updateNotebookAffinity(e, vscode.NotebookControllerAffinity.Preferred)
+        }
```

Comment:
I'm pretty sure that JupyterLab for example prompts the user to select a kernel in that case, maybe we should mimic that as well? Or we could come up with something more clever. Say if there is only one Julia version installed, we use that. If there are multiple, we use the newest (unless that is older than the version in the notebook file). Or something like that... But I think maybe for starters I would probably just not set anything preferred and trigger the UI for kernel selection in this case.